src/protocol/ipa_prf/prf_sharding/mod.rs

```diff
@@ -712,64 +700,62 @@ where
         did_trigger_get_attributed.clone()
     };
 
-    Ok(BitDecomposed::new(
-        ctx.parallel_join(
-            trigger_value
-                .iter()
-                .zip(repeat(zero_out_flag))
-                .enumerate()
-                .map(|(i, (trigger_value_bit, zero_out_flag))| {
-                    let c = ctx.narrow(&BitOpStep::from(i));
-                    async move {
-                        trigger_value_bit
-                            .multiply(&zero_out_flag, c, record_id)
-                            .await
-                    }
-                }),
-        )
-        .await?,
-    ))
+    // expand zero_out_flag to array
+    let zero_out_flag_array = Replicated::<TV>::expand(&zero_out_flag);
+
+    if_else(
+        ctx,
+        record_id,
+        &zero_out_flag_array,
+        trigger_value,
+        &Replicated::<TV>::ZERO,
+    )
+    .await
 }
 
 /// If the `attribution_window_seconds` is not `None`, we calculate the time
 /// difference between the trigger event and the most recent source event, and
 /// returns a secret-shared bit indicating if the trigger event is within the
 /// attribution window.
-async fn is_trigger_event_within_attribution_window<C>(
+async fn is_trigger_event_within_attribution_window<C, TS>(
     ctx: C,
     record_id: RecordId,
     attribution_window_seconds: Option<NonZeroU32>,
-    trigger_event_timestamp: &BitDecomposed<Replicated<Gf2>>,
-    source_event_timestamp: &BitDecomposed<Replicated<Gf2>>,
-) -> Result<Replicated<Gf2>, Error>
+    trigger_event_timestamp: &Replicated<TS>,
+    source_event_timestamp: &Replicated<TS>,
+) -> Result<Replicated<Boolean>, Error>
 where
-    C: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    C: Context,
+    TS: WeakSharedValue,
+    TS: WeakSharedValue + CustomArray<Element = Boolean> + Field,
+    for<'a> &'a Replicated<TS>: IntoIterator<Item = Replicated<Boolean>>,
 {
     if let Some(attribution_window_seconds) = attribution_window_seconds {
-        assert_eq!(trigger_event_timestamp.len(), source_event_timestamp.len());
+        let time_delta_bits = integer_sub(
+            ctx.narrow(&Step::ComputeTimeDelta),
+            record_id,
+            trigger_event_timestamp,
+            source_event_timestamp,
+        )
+        .await?;
 
-        let time_delta_bits = trigger_event_timestamp
-            .sub(
-                ctx.narrow(&Step::ComputeTimeDelta),
-                record_id,
-                source_event_timestamp,
-            )
-            .await?;
+        let mut constant_bits = <TS as WeakSharedValue>::ZERO;
+        for i in 0..<TS as WeakSharedValue>::BITS {
+            let bit = Boolean::from((attribution_window_seconds.get() >> i) & 1 == 1);
+            constant_bits.set(i.try_into().unwrap(), bit);
+        }
```

Comment:
This looks like a `From` implementation or similar conversion function that belongs on TS.